cookbooks/swift/recipes/source.rb

```diff
@@ -117,26 +103,39 @@
   # seems to pull back pbr to 0.11 and break everything; not installing
   # swift-bench's test-requirements is probably better than that
   command "pip install -e ."
-  if not node['full_reprovision']
-    creates "/usr/local/lib/python2.7/dist-packages/swift-bench.egg-link"
-  end
   action :run
 end
 
 execute "python-swift-install" do
   cwd "#{node['source_root']}/swift"
-  command "pip install #{if not node['use_python3'] then '-c py2-constraints.txt' end} -e .[kmip_keymaster] -r test-requirements.txt"
-  if not node['full_reprovision']
-    creates "/usr/local/lib/python2.7/dist-packages/swift.egg-link"
-  end
+  command "pip install -e .[kmip_keymaster] -r test-requirements.txt"
+  action :run
+end
+
+# ignore this, apparently ohai doesn't support explicit python3
+# https://github.com/chef/ohai/blob/main/lib/ohai/plugins/python.rb#L26
+py3_ver_str = shell_out('python3 -c "import sys; print(sys.version)"').stdout
+old_py3 = py3_ver_str.split[0].split('.')[1].to_i < 10
+
+# ubuntu has a patched distutils that does egg-link/editable *installs* to
+# /usr/local/lib/../site-packages; but won't look in that path unless
+# VIRTUAL_ENV=1 (I mean who installs editable python packages as root!?)
+
+# N.B. the "correct" answer on ubuntu is actually "dist-packages"
+site_packages = shell_out('python3 -c "import site; print(site.getsitepackages()[0])"').stdout
+
+# N.B. swift-bench has this same problem; but it goes away when the project
+# grows a pyproject.toml
+# 899958: add pyproject.toml to support pip 23.1 | https://review.opendev.org/c/openstack/swift/+/899958
+execute "legacy-python-swift-install" do
+  cwd "#{node['source_root']}/swift"
+  command "sudo python3 setup.py develop --script-dir /usr/local/bin --install-dir=#{site_packages}"
```

Comment:
FWIW, [Project Testing Interface](https://governance.openstack.org/tc/reference/pti/python.html) is still very much `setup.py`-based.

Reply:
I didn't know that and also I'm not sure how it relates to this hack? I don't have a problem with `python3 setup.py develop` that much; it sounds like `pip install -e .` is actually doing that under the hood because we don't have a `pyproject.toml` which IIUC is a popular thing for modern python packages to do?

https://packaging.python.org/en/latest/guides/writing-pyproject-toml/